### ISSUE TITLE
zotfile fails to rename attachments, when >1 attachments

### ISSUE THREAD

**Issue by maxheld83** [jlegewie/zotfile#283]
It appears that zotfile fails to rename attachments, when there are >1 attachments and when these are imported in bulk.

This is best reproduced by *also* using the Better Bibtex plugin in [this nightly built](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-1.6.96-br667-3750.xpi), *though the problem appears to be on zotfile's end*.

Here's a testcase to show the problem: 
[attachment-test.zip](https://github.com/jlegewie/zotfile/files/1004037/attachment-test.zip)

To reproduce, install the BBT plugin from the above build, current zotfile and then import `attachment-test-new.bib` (never mind the other `*.bib`).

This imports and renames fine, *except* for the record which has *two* attachments; here, the attachments remain un-renamed.

<img width="1339" alt="zotfileerror" src="https://cloud.githubusercontent.com/assets/5372770/26102329/de948a5a-3a34-11e7-8c5a-49ea68e031e3.png">

**Comment by retorquere** [comment]
A new BBT release should be out in 20-30 minutes when [this](https://circleci.com/gh/retorquere/zotero-better-bibtex/3752). Better to wait for that than to use test builds.

**Comment by jlegewie** [comment]
You are talking about the automatic renaming when a new attachment is added to zotero, right? Or the manual renaming?

**Comment by retorquere** [comment]
If I'd have to venture a guess I'd look at timing issues when attachments are added in quick succession (which is the only way I can envision bbt playing a role here).

**Comment by jlegewie** [comment]
Thanks for the report. Not sure when I will get to fixing it though but I will put it on the list for 5.0 final.

**Comment by specter119** [comment]
@jlegewie When I try to manual renaming all items, for attachments > 1, the links will be lost for 2nd and 3rd ... files.

**Comment by jlegewie** [comment]
@maxheld83 Can you try to set `extensions.zotfile.automatic_renaming` to 4 and let me know what happens? You can change this setting under Zotero -> Preferences -> Advanced -> Open about:config.

**Comment by chaoqing** [comment]
@jlegewie The same problem here. Set extensions.zotfile.automatic_renaming to 4 do not solve this problem.

**Comment by aliakhavan89** [comment]
Hi guys! I was wondering whether the issue has been resolved by now? This is because I have the same problem and I did everything mentioned here and other forums but none of them worked.

**Comment by jlegewie** [comment]
I will try to take a look next week but can’t promise.

**Comment by aliakhavan89** [comment]
Thanks! There is also this discussion that may be helpful to better understand the issue:

https://forums.zotero.org/discussion/comment/309449#Comment_309449

**Comment by jlegewie** [comment]
So I looked at this a little more and actually think that automatic renaming works in most cases. Here are four scenarios. 1 and 2 work and I think are the most common workflows. 3 and 4 don't work with zotfile's automatic renaming. I am not sure whether 4 is even possible because Zotero does not trigger an event when retrieving the metadata is done. I will leave this issue open but won't fix 3 anytime soon.

1. Add item to zotero using Zotero Connector (e.g. in FF). Automatic renaming works.
2. Drag and drop one pdf on existing zotero item. Automatic renaming works.
3. Drag and drop multiple pdfs on existing zotero item. Some problems
4. Drag and drop pdf on Zotero (not existing item), Zotero will retrieve the metadata and create a parent item. Automatic does not work.

**Comment by emichael** [comment]
I'm noticing similar behavior to @jlegewie. Just to add one more data point: Zotero Connector (secario 1 above) will only result in automatic renaming when the full item is created by the extension (e.g., when adding from arXiv). When only the PDF is created by Zotero Connector, Zotero metadata etc. and automatic renaming doesn't happen.

**Comment by eulerreich** [comment]
I'd like to voice my support for a fix for (4). It can also be triggered if you are viewing a pdf in the browser and save using Zotero connector. About half of the time for me the pdf never gets renamed because of it.

**Comment by Alexander-Serov** [comment]
> So I looked at this a little more and actually think that automatic renaming works in most cases. Here are four scenarios. 1 and 2 work and I think are the most common workflows. 3 and 4 don't work with zotfile's automatic renaming. I am not sure whether 4 is even possible because Zotero does not trigger an event when retrieving the metadata is done. I will leave this issue open but won't fix 3 anytime soon.
> 
> 1. Add item to zotero using Zotero Connector (e.g. in FF). Automatic renaming works.
> 2. Drag and drop one pdf on existing zotero item. Automatic renaming works.
> 3. Drag and drop multiple pdfs on existing zotero item. Some problems
> 4. Drag and drop pdf on Zotero (not existing item), Zotero will retrieve the metadata and create a parent item. Automatic does not work.

You are right, but `in most cases` formulation is not correct: some people (like me) drag and drop there files in Zotero most of the time. So in most cases, the renaming does not work automatically. ;)

**Comment by harshita-gupta** [comment]
Any updates on issue 4 progress? Really a bummer and frustrating for my most common workflow.

**Comment by Mr-Neffets** [comment]
Any possible updates on this? I'd just like Zotfile to kick in whenever an item is added, whatever the method.
